### PR TITLE
Microsoft/users/raagra/azure tasks stable

### PR DESCRIPTION
**Pull request by rajatagrawal-dev**

### CODE REVIEW

File: Tasks/AzureFileCopy/AzureUtilityRest.ps1 [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+   param([string]$resourceGroupName,
+          [string]$vmName,
+          [string]$name,
+          [object]$endpoint)
+
+    if(-not [string]::IsNullOrEmpty($resourceGroupName) -and -not [string]::IsNullOrEmpty($vmName))
+    {
+        Write-Host (Get-VstsLocString -Key "AFC_GetCustomScriptExtension" -ArgumentList $name, $vmName)
+        $customScriptExtension = Get-AzRmVmCustomScriptExtension $resourceGroupName $vmName $name $endpoint
+        #$customScriptExtension = Get-AzureRmVMCustomScriptExtension -ResourceGroupName $resourceGroupName -VMName $vmName -Name $name -ErrorAction Stop -Verbose
```
Comment (kmkumaran) [inline Tasks/AzureFileCopy/AzureUtilityRest.ps1]: Can you remove this line if not required.

File: Tasks/AzureFileCopy/AzureUtilityRest.ps1 [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+    param([string]$resourceGroupName,
+          [string]$vmName,
+          [string]$name,
+          [object]$endpoint)
+
+    if(-not [string]::IsNullOrEmpty($resourceGroupName) -and -not [string]::IsNullOrEmpty($vmName) -and -not [string]::IsNullOrEmpty($name))
+    {
+        Write-Host (Get-VstsLocString -Key "AFC_RemoveCustomScriptExtension" -ArgumentList $name, $vmName)
+        $response = Remove-AzRmVMCustomScriptExtension $resourceGroupName $vmName $name $endpoint
+        #$response = Remove-AzureRmVMCustomScriptExtension -ResourceGroupName $resourceGroupDetails -VMName $vmName -Name $name -Force -ErrorAction SilentlyContinue -Verbose
```
Comment (kmkumaran) [inline Tasks/AzureFileCopy/AzureUtilityRest.ps1]: Remove this commented line.

File: Tasks/AzureFileCopy/task.loc.json [hunk 3]
```diff
@@ -13,7 +13,7 @@
   "version": {
     "Major": 1,
     "Minor": 0,
-    "Patch": 74
+    "Patch": 75
```
Comment (kmkumaran) [inline Tasks/AzureFileCopy/task.loc.json]: task.json and task.loc.json do not match. Please checkin the latest version post gulp.

File: Tasks/Common/VstsAzureRestHelpers_/VstsAzureRestHelpers_.psm1 [hunk 4]
```diff
@@ -0,0 +1,402 @@
+﻿# Private module-scope variables.
+$script:azureUri = "https://management.core.windows.net"
+$script:azureRmUri = "https://management.azure.com"
```
Comment (kmkumaran) [inline Tasks/Common/VstsAzureRestHelpers_/VstsAzureRestHelpers_.psm1]: Don't hardcode. Can you use this from the endpoint object. We will break support for other Azure Cloud environment types.

File: Tasks/Common/VstsAzureRestHelpers_/VstsAzureRestHelpers_.psm1 [hunk 5]
```diff
@@ -0,0 +1,402 @@
...
+    [CmdletBinding()]
+    param([Parameter(Mandatory=$true)] $endpoint)
+
+    $principalId = $endpoint.Auth.Parameters.ServicePrincipalId
+    $tenantId = $endpoint.Auth.Parameters.TenantId
+    $principalKey = $endpoint.Auth.Parameters.ServicePrincipalKey
+    
+    # Prepare contents for POST
+    $method = "POST"
+    $authUri = "https://login.windows.net/$tenantId/oauth2/token"
```
Comment (kmkumaran) [inline Tasks/Common/VstsAzureRestHelpers_/VstsAzureRestHelpers_.psm1]: Is this URL safe for supporting other Azure Cloud enviornment types?

File: Tasks/AzureFileCopy/task.loc.json (hunk 3 above)
Comment (rajatagrawal-dev) [inline Tasks/AzureFileCopy/task.loc.json]: they are unequal in current master as well. I've just incremented the patch numbers of task.json and task.loc.json by 1 each from the current master.

File: Tasks/AzureFileCopy/AzureFileCopy.ps1 [hunk 6]
```diff
@@ -70,11 +71,12 @@ if ($enableDetailedLoggingString -ne "true")
     $enableDetailedLoggingString = "false"
 }
 
+
```
Comment (mvvsubbu) [inline Tasks/AzureFileCopy/AzureFileCopy.ps1]: Remove unwanted line

File: Tasks/Common/VstsAzureRestHelpers_/VstsAzureRestHelpers_.psm1 [hunk 7]
```diff
@@ -0,0 +1,418 @@
...
+
+# Import the loc strings.
+Import-VstsLocStrings -LiteralPath $PSScriptRoot/module.json
+
+function Get-AzureUri
+{
+    param([object] [Parameter(Mandatory=$true)] $endpoint)
+
+    $url = $endpoint.url
+    if ($url[-1] -eq '/')
```
Comment (pavanadepu2) [inline Tasks/Common/VstsAzureRestHelpers_/VstsAzureRestHelpers_.psm1]: Place a null check here.

File: Tasks/Common/VstsAzureRestHelpers_/VstsAzureRestHelpers_.psm1 [hunk 8]
```diff
@@ -0,0 +1,418 @@
...
+        $accessToken = Get-SpnAccessToken $endpoint
+
+        $resourceGroupDetails = Get-AzRmResourceGroup $resourceGroupName $endpoint
+        $resourceGroupId = $resourceGroupDetails.id
+        $azureRmUri = Get-AzureUri $endpoint
+
+        $method = "POST"
+        $uri = "$azureRmUri$resourceGroupId/providers/Microsoft.Storage/storageAccounts/$storageAccountName/listKeys" + '?api-version=2015-06-15'
+
+        $headers = @{"x-ms-client-request-id"="d5b6a13d-7fa4-43fd-b912-a83a37221815"}
```
Comment (pavanadepu2) [inline Tasks/Common/VstsAzureRestHelpers_/VstsAzureRestHelpers_.psm1]: Is it a well known identity? Does it has an expiry?

File: Tasks/Common/VstsAzureRestHelpers_/VstsAzureRestHelpers_.psm1 [hunk 9]
```diff
@@ -0,0 +1,418 @@
...
+    {
+        $accessToken = Get-SpnAccessToken $endpoint
+        $resourceGroupDetails = Get-AzRmResourceGroup $resourceGroupName $endpoint
+        $resourceGroupId = $resourceGroupDetails.id
+        $azureRmUri = Get-AzureUri $endpoint
+
+        $method="DELETE"
+        $uri = "$azureRmUri$resourceGroupId/providers/Microsoft.Compute/virtualMachines/$vmName/extensions/$Name" + '?api-version=2016-03-30'
+
+        $headers = @{"x-ms-client-request-id"="f6c57f61-2003-4b56-a34c-d8d41a345f2d"}
```
Comment (pavanadepu2) [inline Tasks/Common/VstsAzureRestHelpers_/VstsAzureRestHelpers_.psm1]: Same comment as above, applicable in all the places.

File: Tasks/Common/VstsAzureRestHelpers_/VstsAzureRestHelpers_.psm1 [hunk 10]
```diff
@@ -0,0 +1,418 @@
...
+        {
+            $storageAccountUnformatted=Invoke-RestMethod -Uri $uri -Method $method -Headers $headers
+            Write-Verbose "No Proxy settings"
+        }
+        else
+        {
+            $storageAccountUnformatted=Invoke-RestMethod -Uri $uri -Method $method -Headers $headers -UseDefaultCredentials -Proxy $proxyUri -ProxyUseDefaultCredentials
+            Write-Verbose "Using Proxy settings"
+        }
+
```
Comment (pavanadepu2) [inline Tasks/Common/VstsAzureRestHelpers_/VstsAzureRestHelpers_.psm1]: Can you add a log message here, something like constructing the storage account.

File: Tasks/Common/VstsAzureRestHelpers_/module.json [hunk 11]
```diff
@@ -0,0 +1,7 @@
+{
+    "messages": {
+        "AZ_UnsupportedAuthScheme0": "Unsupported authentication scheme '{0}' for Azure endpoint.",
+        "AZ_BearerTokenFetchFailure": "Not able to fetch token for tenant Id {0}.",
+        "AZ_NoValidResourceIdFound": "No Valid Resource of ServerName : {0} , ServerType : {1} Found for Subscription {2}."
```
Comment (pavanadepu2) [inline Tasks/Common/VstsAzureRestHelpers_/module.json]: Are using AZ_UnsupportedAuthScheme0 and AZ_NoValidResourceIdFound. If not, remove the unused messages.

File: Tasks/AzureFileCopy/Utility.ps1 [hunk 12]
```diff
@@ -94,34 +94,40 @@ function Get-AzureVersionComparison($azureVersion, $compareVersion)
...
+	
+    if(!(Get-AzureVersionComparison -AzureVersion $currentVersion -CompareVersion $AzureVersion103))
     {
-        $azureUtilityRequiredVersion = $azureUtilityVersion100
+		return $azureUtilityVersion100
     }
-    else
+	
+	$endpoint = Get-TypeOfConnection $connectedServiceName
+	if(!(Get-AzureVersionComparison -AzureVersion $currentVersion -CompareVersion $AzureVersion132) -or ($endpoint.Auth.Scheme -eq "UserNamePassword"))
```
Comment (pavanadepu2) [inline Tasks/AzureFileCopy/Utility.ps1]: Make sure you have a null check for $endpoint members.

Comment (pavanadepu2) [inline Tasks/AzureFileCopy/Utility.ps1]: This check won't honor the case where auth scheme is 'Certificate'. Can you reconfirm this scenario while performing the testing.

File: Tasks/DeployAzureResourceGroup/AzureUtilityRest.ps1 [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+        {
+            Write-Verbose "[Azure Resource Manager]Creating resource group $resourceGroupName in $location"
+            $azureResourceGroup = New-AzureRMResourceGroup -Name $resourceGroupName -Location $location -Verbose -ErrorAction Stop
+            Write-Host (Get-VstsLocString -Key "ARG_CreatedResourceGroup" -ArgumentList $resourceGroupName)
+        }
+        return $azureResourceGroup
+    }
+}
+
+function Get-AzureMachineCustomScriptExtension
```
Comment (pavanadepu2) [inline Tasks/DeployAzureResourceGroup/AzureUtilityRest.ps1]: Is the response json of these calls matching with the azure ps cmdlet responses?

File: Tasks/Common/VstsAzureRestHelpers_/VstsAzureRestHelpers_.psm1 [hunk 14]
```diff
@@ -108,27 +135,40 @@ function Get-SpnAccessToken {
...
+        {
+            Write-Verbose "Using Proxy settings"
+            $accessToken = Invoke-RestMethod -Uri $authUri -Method $method -Body $body -ContentType $script:formContentType -UseDefaultCredentials -Proxy $proxyUri -ProxyUseDefaultCredentials
+            return $accessToken
+        }
     }
     catch
     {
-        throw (Get-VstsLocString -Key AZ_SpnAccessTokenFetchFailure -ArgumentList $tenantId)
+        throw (Get-VstsLocString -Key AZ_BearerTokenFetchFailure -ArgumentList $tenantId)
```
Comment (kmkumaran) [inline Tasks/Common/VstsAzureRestHelpers_/VstsAzureRestHelpers_.psm1]: Where is this resource coming from? I don't see it in the module.json file.

**Pull request merged**